File: reversing_linked_list.py
class node: #Node class
    def __init__(self,data=None,next=None):
        self.data= data
        self.next = next

class linked_list: #linked list class
    def __init__(self,head=None):
        self.head = head


linked_list = linked_list() #a new linked list object

# inputting data
linked_list.head = node(data=input())
s = linked_list.head

for i in range(4):
    s.next = node(data=input())
    s = s.next

s = linked_list.head

print('------------------------pre-reverse--------------------------------')
while 1:
    print(s.data)
    s= s.next
    if not s:
        break

first = linked_list.head #reference to the initial head of the list
second = first.next #reference to the initial second element of the list
third = second.next #reference to the initial third element of the list

while second.next is not None: #loop till second.next is not None
    first.next = third #set first.next to third
    second.next = linked_list.head #set second.next to the current head value
    linked_list.head = second #change the current head value to second

    second = third #change second to third
    third = third.next #change third to third.next
#when the loop exits do the last swap
first.next = third #change first.next to third
second.next = linked_list.head #set second.next to current head
linked_list.head = second #set linked_list.head as second

s= linked_list.head
print('------------------------post-reverse--------------------------------')
while 1:
    print(s.data)
    s= s.next
    if not s:
        break

File: reversing_linked_list.cpp
#include<iostream>
using namespace std;
class linked{
public:
    int data;
    linked *next;
};

void fill_list(linked *node){

for (int i =0; i <10; i ++){
    cin>>node->data;
    linked *new_node = new linked();
    node->next = new_node;
    node = node->next;
}

}

void print(linked *node){
    cout<<node->data<<"->";
    node = node->next;
    while(true){
        cout<<node->data<<"->";
        node=node->next;
        if(!node->next){
            break;
        }
    }

}
void reverse_list(linked **node){

    linked *first= *node; // pointer first points to the pointer of node
    linked *second = first->next; // pointer second points to first->next
    linked *third = second->next; // pointer third points to second->next



    while(second->next){ //loop till second->next is not null
        first->next = third; // first-> next now points to third
        second->next = *node; // second->next now points to the head node
        *node=second; // pointer node which is the pointer to head now points to second
        second = third; // second is now third
        third = third->next; // third is now third->next

    }

}

int main(){

linked *node = new linked();

fill_list(node);
cout<<"Before Reversing\n\n";
print(node);
reverse_list(&node);
cout<<"\n\nAfter Reversing\n\n";
print(node);
cout<<"\n";

}
